### PR TITLE
Add version flag to host agent

### LINKED ISSUE #278: Add --version flag to host agent binary
It will be useful to provide a `--version` flag in the host-agent binary as we will have multiple agent releases so that the user can know the version being used. The output could consist of information in the below format

```
$ byoh-hostagent-linux-amd64 --version
Version: v0.1.0_alpha.2
BuildDate: 
...
```

### PR DESCRIPTION
**Pull request by raunaksingwi**
**What this PR does / why we need it**:
Add version flag to the host agent so that the user can know the version being used.
The output shows the version of the binary and the time it was built.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #278

**Additional information**
Used ldflags to embed the version number and time at build time in the binary.

The version number has to be passed as a variable to the Makefile. It will be passed by the GH action while the build is triggered.
When the version is not passed, it takes the default as "dev"

**Special notes for your reviewer**

### CODE REVIEW

File: agent/main.go [hunk 1]
```diff
@@ -88,8 +91,15 @@ func main() {
 	flag.StringVar(&metricsbindaddress, "metricsbindaddress", ":8080", "metricsbindaddress is the TCP address that the controller should bind to for serving prometheus metrics.It can be set to \"0\" to disable the metrics serving")
 	flag.StringVar(&downloadpath, "downloadpath", "/var/lib/byoh/bundles", "File System path to keep the downloads")
 	flag.BoolVar(&skipInstallation, "skip-installation", false, "If you want to skip installation of the kubernetes component binaries")
+	flag.BoolVar(&printVersion, "version", false, "Print the version of the agent")
 	klog.InitFlags(nil)
 	flag.Parse()
+
+	if printVersion {
```
Comment (anusha94) [inline agent/main.go:98]: Can we add tests to validate this?
You could probably add a new Context in this file [host_agent_test.go](https://github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/blob/main/agent/host_agent_test.go)

File: agent/main.go [hunk 2]
```diff
@@ -78,6 +78,9 @@ var (
 	metricsbindaddress string
 	downloadpath       string
 	skipInstallation   bool
+	printVersion       bool
+	BuildVersion       string = ""
```
Comment (dharmjit) [inline agent/main.go:82]: Nitpick: Instead of defining a default value in Makefile can it be initialized to `dev` here.

File: Makefile [hunk 3]
```diff
@@ -154,7 +155,7 @@ kustomize: ## Download kustomize locally if necessary.
 	$(call go-get-tool,$(KUSTOMIZE),sigs.k8s.io/kustomize/kustomize/v3@v3.9.1)
 
 host-agent-binaries: ## Builds the binaries for the host-agent
-	RELEASE_BINARY=./byoh-hostagent GOOS=linux GOARCH=amd64 GOLDFLAGS="$(LDFLAGS) $(STATIC)" HOST_AGENT_DIR=./$(HOST_AGENT_DIR) $(MAKE) host-agent-binary
+	RELEASE_BINARY=./byoh-hostagent GOOS=linux GOARCH=amd64 GOLDFLAGS="$(LDFLAGS) $(STATIC) -X 'main.BuildVersion=${AGENT_BINARY_VERSION}' -X 'main.BuildTime=$(shell date)'" HOST_AGENT_DIR=./$(HOST_AGENT_DIR) $(MAKE) host-agent-binary
```
Comment (dharmjit) [inline Makefile:158]: Nitpick: Is it possible to convert the `shell date` format to `RFC3339`.

**Review by dharmjit — COMMENTED** [review]
some basic nits only. 

IMO it is useful to keep `GoVersion` and `Platform` as well. `kubectl` version output below. Thoughts @anusha94 @jamiemonserrate 
```
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.1", GitCommit:"86ec240af8cbd1b60bcc4c03c20da9b98005b92e", GitTreeState:"clean", BuildDate:"2021-12-16T11:41:01Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
```

File: agent/main.go (hunk 2 above)
Comment (raunaksingwi) [reply agent/main.go:82]: The Makefile has to pass some value for this variable as it is part of the build command ldflags. And that value will be overwritten to this variable, even if it has a default value.

The alternative will be to have a conditional block in Makefile where if the BuildVersion isn't set in Makefile we run another build command which doesn't pass this value, but that approach would be less neat.

Comment (dharmjit) [reply agent/main.go:82]: > And that value will be overwritten to this variable, even if it has a default value
yup makes sense. Thanks for correcting me.

File: Makefile (hunk 3 above)
Comment (raunaksingwi) [reply Makefile:158]: Updated the output:

When the version is passed to Makefile:

```
bin/byoh-hostagent-linux-amd64 --version

agent version: version.Info{Major:"0", Minor:"1", Patch:"0", BuildDate:"2022-01-12T09:19:25Z", GoVersion:"go1.16.6", Platform:"linux/amd64", Compiler:"gc"}
```

When the version is not passed as default "dev" is used

```
bin/byoh-hostagent-linux-amd64 --version

agent version: version.Info{Major:"dev", Minor:"", Patch:"", BuildDate:"2022-01-12T09:19:25Z", GoVersion:"go1.16.6", Platform:"linux/amd64", Compiler:"gc"}
```

File: agent/host_agent_test.go [hunk 4]
```diff
@@ -188,4 +192,45 @@ var _ = Describe("Agent", func() {
...
+				BuildDate: version.BuildDate,
+				GoVersion: runtime.Version(),
+				Compiler:  runtime.Compiler,
+				Platform:  fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH),
+			}
+			expected := fmt.Sprintf("byoh-hostagent version: %#v\n", expectedStruct)
+			out, err := exec.Command(tmpHostAgentBinary, "--version").Output()
+			output := string(out)
+			fmt.Fprintf(GinkgoWriter, "version.Version %s", version.Version)
+			Expect(err).NotTo(HaveOccurred())
```
Comment (anusha94) [inline agent/host_agent_test.go:232]: Can you move this a couple lines up where it gets the `err` value?

File: agent/host_agent_test.go [hunk 5]
```diff
@@ -188,4 +192,45 @@ var _ = Describe("Agent", func() {
...
+				Patch:     "3",
+				BuildDate: version.BuildDate,
+				GoVersion: runtime.Version(),
+				Compiler:  runtime.Compiler,
+				Platform:  fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH),
+			}
+			expected := fmt.Sprintf("byoh-hostagent version: %#v\n", expectedStruct)
+			out, err := exec.Command(tmpHostAgentBinary, "--version").Output()
+			output := string(out)
+			fmt.Fprintf(GinkgoWriter, "version.Version %s", version.Version)
```
Comment (anusha94) [inline agent/host_agent_test.go:231]: I don't think this logging is required. If the test fails, ginkgo will anyway log it.

File: agent/installer/internal/algo/shell_step.go [hunk 6]
```diff
@@ -1,81 +1,81 @@
...
-		return err
-	}
-
-	if len(stdOut.String()) > 0 {
-		s.OutputBuilder.Out(stdOut.String())
-	}
-
-	return nil
-}
+// Copyright 2021 VMware, Inc. All Rights Reserved.
```
Comment (anusha94) [inline agent/installer/internal/algo/shell_step.go:1]: I'm not sure why this file is changed. But I'm hoping there are no changes

File: agent/version/version_test.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+					Compiler:  runtime.Compiler,
+					Platform:  fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH),
+				}
+				Expect(version.Get()).Should(Equal(expected))
+			})
+		})
+
+		Context("When version is set as git tag", func() {
+			BeforeEach(func() {
+				version.Version = "1.2.3"
```
Comment (anusha94) [inline agent/version/version_test.go:86]: The git tag will be `v1.2.3`
Does it handle this?

**Comment by anusha94** [comment]
Also, please squash all commits. All the merge commits seem to have been included

File: agent/installer/internal/algo/shell_step.go (hunk 6 above)
Comment (raunaksingwi) [reply agent/installer/internal/algo/shell_step.go:1]: This, and many other files got in as part of rebasing with main.
These must be files that diverged.

File: agent/host_agent_test.go (hunk 4 above)
Comment (raunaksingwi) [reply agent/host_agent_test.go:229]: fixed

File: agent/host_agent_test.go (hunk 5 above)
Comment (raunaksingwi) [reply agent/host_agent_test.go]: removed

**Pull request closed without merging**